### PR TITLE
Feat/upgrade node ts

### LINKED ISSUE #164: Upgrade typescript to support Array toReversed
I'm using a destructing array to prevent mutation to the original list, since we can't use `es2023` stuff now. 

_Originally posted by @laporchen in https://github.com/Normal-OJ/new-front-end/pull/158#discussion_r1527427053_

### PR DESCRIPTION
**Pull request by uier**
fix #164 #151

### CODE REVIEW

**Comment by uier** [comment]
oh, seems the failing test can be reproduced on my local machine.

**Review by laporchen — APPROVED** [review]
LGTM. No breaking changes from 16 -> 20 it seems

File: .github/workflows/lint-ci.yaml [hunk 1]
```diff
@@ -15,10 +15,10 @@ jobs:
       - uses: pnpm/action-setup@v2
         with:
           version: 8.6
-      - name: Use Node.js v16.x
+      - name: Use Node.js v20.x
         uses: actions/setup-node@v3
```
Comment (Bogay) [inline .github/workflows/lint-ci.yaml:19]: 要不要順便升 v4

File: .github/workflows/playwright.yml [hunk 2]
```diff
@@ -26,10 +26,10 @@ jobs:
       - uses: pnpm/action-setup@v2
         with:
           version: 8.6
-      - name: Use Node.js v16.x
+      - name: Use Node.js v20.x
         uses: actions/setup-node@v3
```
Comment (Bogay) [inline .github/workflows/playwright.yml:30]: 這邊也是

**Review by Bogay — APPROVED** [review]

File: .github/workflows/lint-ci.yaml (hunk 1 above)
Comment (uier) [reply .github/workflows/lint-ci.yaml:19]: 差在哪

Comment (Bogay) [reply .github/workflows/lint-ci.yaml:19]: 大概就 runtime 版本還有一些 dependency 升一升

https://github.com/actions/setup-node/releases/tag/v4.0.0

**Comment by uier** [comment]
Upgrade node ver on Cloudflare (by changing env) before merging this into main

**Pull request merged**